Only use the host if its Resource State is Enabled.

mike-tutkowski opened the pull request:
## Description
Update the StorageSystemSnapshotStrategy and StorageSystemDataMotionStrategy classes to only leverage a host when its Resource State is Enabled.

https://issues.apache.org/jira/browse/CLOUDSTACK-10345

## Types of changes
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

## How Has This Been Tested?

Checked to see if a host in maintenance mode is avoided

## Checklist:
- [x] I have read the [CONTRIBUTING](https://github.com/apache/cloudstack/blob/master/CONTRIBUTING.md) document.
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

<!-- The following will kick a packaging job, remove it as applicable -->
@blueorangutan package

blueorangutan commented:
@mike-tutkowski a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

nitin-maharana commented on engine/storage/snapshot/src/main/java/org/apache/cloudstack/storage/snapshot/StorageSystemSnapshotStrategy.java:
```diff
@@ -474,7 +473,7 @@ else if (volumeInfo.getFormat() == ImageFormat.OVA || volumeInfo.getFormat() ==
                 vmSnapshot = takeHypervisorSnapshot(volumeInfo);
             }
-            catch (ResourceAllocationException ex) {
+            catch (Exception ex) {
```
@mike-tutkowski, Thanks for the fix. I think it would be better to capture the required exceptions with multiple catches instead of the parent Exception class. We can always put this at the end.

nitin-maharana commented on engine/storage/snapshot/src/main/java/org/apache/cloudstack/storage/snapshot/StorageSystemSnapshotStrategy.java:
```diff
@@ -484,7 +483,7 @@ else if (volumeInfo.getFormat() == ImageFormat.OVA || volumeInfo.getFormat() ==
 
         SnapshotResult result = null;
-        SnapshotInfo snapshotOnPrimary = null;
+        SnapshotInfo snapshotOnPrimary;
```
I guess the removal of this may create NPE.

DaanHoogland replied:
how is this, @nitin-maharana?

mike-tutkowski replied:
This is an IntelliJ-recommended change. The IDE pointed out (correctly) that we did not need to assign null on this line because snapshotOnPrimary is later assigned a value regardless of the path the code takes.

nitin-maharana replied:
@DaanHoogland, My bad, wrong interpretation!! As @mike-tutkowski explained, that's redundant.  Thanks!!

blueorangutan commented:
Packaging result: ✔centos6 ✔centos7 ✔debian. JID-1834

rhtyd commented:
@mike-tutkowski bugfixes can be considered for 4.11 branch, if you think this is useful for 4.11/tls please change the base-branch to 4.11 (and rebase your branch against 4.11). Please also put this PR against a milestone, 4.11.1.0, or 4.12.0.0 as applicable.

mike-tutkowski commented:
@rhtyd Sounds good, Rohit. I can change the base branch to 4.11 and rebase. I'll do that after I implement @nitin-maharana's suggestion around exception handling.

mike-tutkowski commented:
@rhtyd @nitin-maharana @DaanHoogland I have rebased the code on top of 4.11 and changed the base branch of the PR. Also, @nitin-maharana: I was able to entirely remove that try/catch block.

DaanHoogland commented:
@blueorangutan package

blueorangutan commented:
@DaanHoogland a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✔centos6 ✔centos7 ✔debian. JID-1847

DaanHoogland commented:
@blueorangutan test

blueorangutan commented:
@DaanHoogland a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

blueorangutan commented:
<b>Trillian test result (tid-2435)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 92219 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr2512-t2435-kvm-centos7.zip
Intermitten failure detected: /marvin/tests/smoke/test_privategw_acl.py
Intermitten failure detected: /marvin/tests/smoke/test_routers.py
Intermitten failure detected: /marvin/tests/smoke/test_vpc_vpn.py
Smoke tests completed. 64 look OK, 3 have error(s)
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_02_vpc_privategw_static_routes | `Failure` | 150.61 | test_privategw_acl.py
test_03_vpc_privategw_restart_vpc_cleanup | `Failure` | 150.58 | test_privategw_acl.py
test_04_rvpc_privategw_static_routes | `Failure` | 272.80 | test_privategw_acl.py
test_04_restart_network_wo_cleanup | `Failure` | 2.79 | test_routers.py
test_01_vpc_site2site_vpn | `Failure` | 220.59 | test_vpc_vpn.py

mike-tutkowski commented:
Just a quick comment on the five test failures: It seems unlikely that this PR caused them. Those are all networking related and this PR is mainly concerned with managed storage.

blueorangutan commented:
@mike-tutkowski a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

mike-tutkowski commented:
@DaanHoogland Quick question: Is it part of our process to forward merge these PRs to master or do I need to open a new PR for that? Thanks!

rafaelweingartner commented:
I was going to mention that just now. No new PR should be required. We should merge forward.
There is also another PR that @DaanHoogland merged yesterday that has not been merged forward.

DaanHoogland commented:
I think we are at more than ten now :s , gents

DaanHoogland commented:
but I think it doesn't hurt to gather a few before we merge forward.

rafaelweingartner commented:
Really?! I normally follow PRs merge, but I only notice the one yesterday and this one.
Well, I guest that as long as we merge them forward, it is ok to do the merges in batch.
